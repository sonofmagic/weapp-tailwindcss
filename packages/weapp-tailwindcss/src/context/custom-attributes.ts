import type {
  ICustomAttributes,
  ICustomAttributesEntities,
  ItemOrItemArray,
} from '@/types'
import { isMap } from '@/utils'

export function toCustomAttributesEntities(customAttributes: ICustomAttributes): ICustomAttributesEntities {
  if (isMap(customAttributes)) {
    return [
      ...(customAttributes as Exclude<ICustomAttributes, Record<string, ItemOrItemArray<string | RegExp>>>).entries(),
    ]
  }

  return Object.entries(customAttributes)
}
