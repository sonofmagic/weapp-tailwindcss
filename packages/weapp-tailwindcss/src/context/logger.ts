import { logger } from '@weapp-tailwindcss/logger'

// https://www.npmjs.com/package/consola
const loggerLevelMap: Record<'info' | 'warn' | 'error' | 'silent', number> = {
  error: 0,
  warn: 1,
  info: 3,
  silent: -999,
}

export type LoggerLevel = keyof typeof loggerLevelMap

export function applyLoggerLevel(logLevel?: LoggerLevel): void {
  logger.level = loggerLevelMap[logLevel ?? 'info'] ?? loggerLevelMap.info
}
