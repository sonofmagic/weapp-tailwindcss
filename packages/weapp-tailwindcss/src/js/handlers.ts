// import type { StringLiteral, TemplateElement, Comment } from '@babel/types'
import type MagicString from 'magic-string'
import type { IJsHandlerOptions } from '../types'
import { replaceWxml } from '../wxml/shared'
import { escapeStringRegexp } from '../reg'
import { splitCode } from '../extractors/split'
import { decodeUnicode2, jsStringEscape } from '../escape'
// https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/String

interface ReplaceNode {
  leadingComments?: { value: string }[] | null | undefined
  start?: number | null
  end?: number | null
}

export function toUnicodeEscapedString(str: string) {
  return str.split('').map((char) => {
    const code = char.charCodeAt(0)
    if (code <= 31) {
      const hexCode = code.toString(16).padStart(4, '0')
      return `\\u${hexCode}`
    }
    return char
  }).join('')
}

export function toUnicodeEscapedString2(str: string) {
  return escape(str).replace(/%u/g, '\\u').replace(/%/g, '\\u00')
}

export function replaceHandleValue(
  str: string,
  node: ReplaceNode,
  options: IJsHandlerOptions,
  ms: MagicString,
  offset = 0,
) {
  const {
    classNameSet: set,
    escapeMap,
    mangleContext: ctx,
    needEscaped = false,
    jsPreserveClass,
    arbitraryValues,
    always,
    unescapeUnicode,
  } = options

  const allowDoubleQuotes = arbitraryValues?.allowDoubleQuotes

  let rawStr = str
  let needUpdate = false
  if (unescapeUnicode && rawStr.includes('\\u')) {
    rawStr = decodeUnicode2(rawStr)
  }
  const arr = splitCode(rawStr, allowDoubleQuotes)
  for (const v of arr) {
    if (always || (set && set.has(v) && !jsPreserveClass?.(v))) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag
          = node.leadingComments.findIndex(x => x.value.includes('weapp-tw') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        if (ctx) {
          rawStr = ctx.jsHandler(rawStr)
        }

        rawStr = rawStr.replace(
          new RegExp(escapeStringRegexp(v)),
          replaceWxml(v, {
            escapeMap,
          }),
        )
        needUpdate = true
      }
    }
  }

  if (needUpdate && typeof node.start === 'number' && typeof node.end === 'number') {
    const start = node.start + offset
    const end = node.end - offset

    if (start < end && str !== rawStr) {
      const content = needEscaped ? jsStringEscape(rawStr) : rawStr
      ms.update(start, end, content)
    }
  }

  return rawStr
}
