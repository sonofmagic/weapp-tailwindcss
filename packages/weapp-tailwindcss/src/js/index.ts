import type { CreateJsHandlerOptions, IJsHandlerOptions, JsHandler } from '../types'
import { defuOverrideArray } from '../utils'
import { jsHandler } from './babel'

export {
  jsHandler,
}

export function createJsHandler(options: CreateJsHandlerOptions): JsHandler {
  // Persist immutable options so repeated invocations only supply per-call overrides.
  const {
    mangleContext,
    arbitraryValues,
    escapeMap,
    jsPreserveClass,
    generateMap,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
  } = options

  function handler(rawSource: string, classNameSet?: Set<string>, options?: CreateJsHandlerOptions) {
    const overrideOptions = (options ?? {}) as IJsHandlerOptions
    const resolvedOptions = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(
      {
        ...overrideOptions,
        classNameSet,
      },
      {
        classNameSet,
        escapeMap,
        arbitraryValues,
        mangleContext,
        jsPreserveClass,
        generateMap,
        babelParserOptions,
        ignoreCallExpressionIdentifiers,
        ignoreTaggedTemplateExpressionIdentifiers,
        uniAppX,
      },
    )

    return jsHandler(rawSource, resolvedOptions)
  }

  return handler
}
