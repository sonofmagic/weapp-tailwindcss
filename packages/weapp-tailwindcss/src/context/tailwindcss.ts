import type { TailwindcssPatchOptions } from 'tailwindcss-patch'
import type { InternalUserDefinedOptions } from '@/types'
import { createRequire } from 'node:module'
import path from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import { logger } from '@weapp-tailwindcss/logger'
import { createTailwindcssPatcher } from '@/tailwindcss'
import { defuOverrideArray } from '@/utils'

const ENV_BASEDIR_KEYS = [
  'WEAPP_TAILWINDCSS_BASEDIR',
  'WEAPP_TAILWINDCSS_BASE_DIR',
  'TAILWINDCSS_BASEDIR',
  'TAILWINDCSS_BASE_DIR',
  'UNI_INPUT_DIR',
  'UNI_INPUT_ROOT',
  'UNI_CLI_ROOT',
  'UNI_APP_INPUT_DIR',
  'INIT_CWD',
  'PWD',
] as const

type EnvBasedirKey = typeof ENV_BASEDIR_KEYS[number]

const GENERIC_ENV_BASEDIR_KEYS = new Set<EnvBasedirKey>(['INIT_CWD', 'PWD'])

interface EnvBasedirResult {
  key: EnvBasedirKey
  value: string
}

function pickEnvBasedir(): EnvBasedirResult | undefined {
  for (const key of ENV_BASEDIR_KEYS) {
    const value = process.env[key]
    if (value && path.isAbsolute(value)) {
      return { key, value }
    }
  }
  return undefined
}

function pickPackageEnvBasedir(): string | undefined {
  const packageJsonPath = process.env.npm_package_json
  if (packageJsonPath) {
    const packageDir = path.dirname(packageJsonPath)
    if (packageDir && path.isAbsolute(packageDir)) {
      return packageDir
    }
  }
  const localPrefix = process.env.npm_config_local_prefix
  if (localPrefix && path.isAbsolute(localPrefix)) {
    return localPrefix
  }
  return undefined
}

function detectCallerBasedir(): string | undefined {
  const stack = new Error('resolveTailwindcssBasedir stack probe').stack
  if (!stack) {
    return undefined
  }
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('caller stack: %s', stack)
  }
  const lines = stack.split('\n')
  for (const line of lines) {
    const match = line.match(/\(([^)]+)\)/u) ?? line.match(/at\s+(\S.*)$/u)
    const location = match?.[1]
    if (!location) {
      continue
    }
    let filePath = location
    if (filePath.startsWith('file://')) {
      try {
        filePath = fileURLToPath(filePath)
      }
      catch {
        continue
      }
    }
    const [candidate] = filePath.split(':')
    const resolvedPath = path.isAbsolute(filePath) ? filePath : candidate
    if (!path.isAbsolute(resolvedPath)) {
      continue
    }
    if (resolvedPath.includes('node_modules') && resolvedPath.includes('weapp-tailwindcss')) {
      continue
    }
    try {
      return path.dirname(resolvedPath)
    }
    catch {
      continue
    }
  }
  return undefined
}

export function resolveTailwindcssBasedir(basedir?: string) {
  const envBasedirResult = pickEnvBasedir()
  const envBasedir = envBasedirResult?.value
  const envBasedirKey = envBasedirResult?.key
  const envBasedirIsGeneric = envBasedirKey ? GENERIC_ENV_BASEDIR_KEYS.has(envBasedirKey) : false
  const packageEnvBasedir = pickPackageEnvBasedir()
  const shouldDetectCaller = !envBasedir || envBasedirIsGeneric
  const callerBasedir = shouldDetectCaller ? detectCallerBasedir() : undefined
  const anchor = envBasedir ?? packageEnvBasedir ?? callerBasedir ?? process.cwd()
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('resolveTailwindcssBasedir anchor %O', {
      basedir,
      envBasedir,
      envBasedirKey,
      envBasedirIsGeneric,
      packageEnvBasedir,
      callerBasedir,
      npm_package_json: process.env.npm_package_json,
      cwd: process.cwd(),
      anchor,
    })
  }

  if (basedir && basedir.trim().length > 0) {
    if (path.isAbsolute(basedir)) {
      return path.normalize(basedir)
    }
    return path.resolve(anchor, basedir)
  }

  if (envBasedir && !envBasedirIsGeneric) {
    return path.normalize(envBasedir)
  }

  if (packageEnvBasedir) {
    return path.normalize(packageEnvBasedir)
  }

  if (callerBasedir) {
    const normalizedCaller = path.normalize(callerBasedir)
    const librarySegment = `${path.sep}weapp-tailwindcss${path.sep}`
    if (!normalizedCaller.includes(librarySegment)) {
      return normalizedCaller
    }
  }

  if (envBasedir) {
    return path.normalize(envBasedir)
  }

  const packageName = process.env.PNPM_PACKAGE_NAME
  if (packageName) {
    try {
      const anchorRequire = createRequire(path.join(anchor, '__resolve_tailwindcss_basedir__.cjs'))
      const packageJsonPath = anchorRequire.resolve(`${packageName}/package.json`)
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('package basedir resolved from PNPM_PACKAGE_NAME: %s', packageJsonPath)
      }
      return path.normalize(path.dirname(packageJsonPath))
    }
    catch {
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('failed to resolve package json for %s', packageName)
      }
      // ignore
    }
  }

  return path.normalize(process.cwd())
}

export function createTailwindcssPatcherFromContext(ctx: InternalUserDefinedOptions) {
  const {
    tailwindcssBasedir,
    supportCustomLengthUnitsPatch,
    tailwindcss,
    tailwindcssPatcherOptions,
    cssEntries,
    appType,
  } = ctx

  const resolvedTailwindcssBasedir = resolveTailwindcssBasedir(tailwindcssBasedir)
  ctx.tailwindcssBasedir = resolvedTailwindcssBasedir
  logger.debug('tailwindcss basedir resolved: %s', resolvedTailwindcssBasedir)

  type TailwindUserOptions = NonNullable<TailwindcssPatchOptions['tailwind']>

  const defaultTailwindcssConfig: TailwindUserOptions = {
    cwd: resolvedTailwindcssBasedir,
    v2: {
      cwd: resolvedTailwindcssBasedir,
    },
    v3: {
      cwd: resolvedTailwindcssBasedir,
    },
    v4: {
      base: resolvedTailwindcssBasedir,
      cssEntries,
    },
  }

  if (cssEntries?.length && (tailwindcss == null || tailwindcss.version == null)) {
    defaultTailwindcssConfig.version = 4
  }

  return createTailwindcssPatcher(
    {
      basedir: resolvedTailwindcssBasedir,
      cacheDir: appType === 'mpx' ? 'node_modules/tailwindcss-patch/.cache' : undefined,
      supportCustomLengthUnitsPatch: supportCustomLengthUnitsPatch ?? true,
      tailwindcss: defuOverrideArray<TailwindUserOptions, TailwindUserOptions[]>(
        (tailwindcss ?? {}) as TailwindUserOptions,
        defaultTailwindcssConfig,
      ),
      tailwindcssPatcherOptions,
    },
  )
}
