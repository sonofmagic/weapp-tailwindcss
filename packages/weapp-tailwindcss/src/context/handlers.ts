import type { IMangleScopeContext } from '@weapp-tailwindcss/mangle'
import type { ICustomAttributesEntities, InternalUserDefinedOptions } from '@/types'
import { createStyleHandler } from '@weapp-tailwindcss/postcss'
import { createJsHandler } from '@/js'
import { createTemplateHandler } from '@/wxml'

export function createHandlersFromContext(
  ctx: InternalUserDefinedOptions,
  mangleContext: IMangleScopeContext,
  customAttributesEntities: ICustomAttributesEntities,
  cssCalcOptions: InternalUserDefinedOptions['cssCalc'],
) {
  const {
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    px2rpx,
    arbitraryValues,
    jsPreserveClass,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    inlineWxs,
    disabledDefaultTemplateHandler,
  } = ctx

  const styleHandler = createStyleHandler({
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    mangleContext,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    cssCalc: cssCalcOptions,
    px2rpx,
  })

  const jsHandler = createJsHandler({
    escapeMap,
    mangleContext,
    arbitraryValues,
    jsPreserveClass,
    generateMap: true,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
  })

  const templateHandler = createTemplateHandler({
    customAttributesEntities,
    escapeMap,
    mangleContext,
    inlineWxs,
    jsHandler,
    disabledDefaultTemplateHandler,
  })

  return {
    styleHandler,
    jsHandler,
    templateHandler,
  }
}
