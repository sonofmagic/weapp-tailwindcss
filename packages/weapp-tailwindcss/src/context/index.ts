import type { InternalUserDefinedOptions, UserDefinedOptions } from '@/types'
import { logger, pc } from '@weapp-tailwindcss/logger'
import { useMangleStore } from '@weapp-tailwindcss/mangle'
import { initializeCache } from '@/cache'
import { getDefaultOptions } from '@/defaults'
import { defuOverrideArray } from '@/utils'
import { toCustomAttributesEntities } from './custom-attributes'
import { createHandlersFromContext } from './handlers'
import { applyLoggerLevel } from './logger'
import { createTailwindcssPatcherFromContext } from './tailwindcss'

/**
 * 获取用户定义选项的内部表示，并初始化相关的处理程序和补丁。
 * @param opts - 用户定义的选项，可选。
 * @returns 返回一个包含内部用户定义选项的对象，包括样式、JS和模板处理程序，以及Tailwind CSS补丁。
 */
export function getCompilerContext(opts?: UserDefinedOptions): InternalUserDefinedOptions {
  const ctx = defuOverrideArray<InternalUserDefinedOptions, Partial<InternalUserDefinedOptions>[]>(
    opts as InternalUserDefinedOptions,
    getDefaultOptions() as InternalUserDefinedOptions,
    {},
  )

  ctx.escapeMap = ctx.customReplaceDictionary

  applyLoggerLevel(ctx.logLevel)

  const twPatcher = createTailwindcssPatcherFromContext(ctx)

  if (twPatcher.packageInfo?.version) {
    logger.success(`当前使用 ${pc.cyanBright('Tailwind CSS')} 版本为: ${pc.underline(pc.bold(pc.green(twPatcher.packageInfo.version)))}`)
  }
  else {
    logger.warn(`${pc.cyanBright('Tailwind CSS')} 未安装，已跳过版本检测与补丁应用。`)
  }

  const cssCalcOptions = ctx.cssCalc ?? twPatcher.majorVersion === 4

  const customAttributesEntities = toCustomAttributesEntities(ctx.customAttributes)

  const { initMangle, mangleContext, setMangleRuntimeSet } = useMangleStore()
  initMangle(ctx.mangle)

  const { styleHandler, jsHandler, templateHandler } = createHandlersFromContext(
    ctx,
    mangleContext,
    customAttributesEntities,
    cssCalcOptions,
  )

  ctx.styleHandler = styleHandler
  ctx.jsHandler = jsHandler
  ctx.templateHandler = templateHandler

  ctx.setMangleRuntimeSet = setMangleRuntimeSet
  ctx.cache = initializeCache(ctx.cache)
  ctx.twPatcher = twPatcher
  return ctx
}
