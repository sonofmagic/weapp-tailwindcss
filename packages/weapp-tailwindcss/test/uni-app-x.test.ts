import { getCompilerContext } from '@/context'
import { createJsHandler } from '@/js'
import { createGetCase, fixturesRootPath } from './util'

const getCase = createGetCase(fixturesRootPath)

describe('uni-app-x', () => {
  it('uvue.ts case 0', async () => {
    const jsHandler = createJsHandler({
      babelParserOptions: {
        sourceType: 'unambiguous',
        plugins: [
          'typescript',
        ],
      },
    })
    const set = new Set<string>()
    set.add('mt-[32.43rpx]')
    set.add('bg-[#322323]')
    set.add('text-[#844343]')
    const { code } = jsHandler(
      await getCase('uni-app-x/index.uvue.ts'),
      set,
      {

      },
    )
    expect(code).toMatchSnapshot()
  })

  it('uvue.ts case 1', async () => {
    const { jsHandler, styleHandler } = getCompilerContext({
      uniAppX: true,
    })
    const set = new Set<string>()
    set.add('mt-[32.43rpx]')
    set.add('bg-[#322323]')
    set.add('text-[#844343]')
    const { code } = jsHandler(
      await getCase('uni-app-x/index.uvue.ts'),
      set,
      {
        babelParserOptions: {
          sourceType: 'unambiguous',
          plugins: [
            'typescript',
          ],
        },
      },
    )
    expect(code).toMatchSnapshot()

    const { css } = await styleHandler(
      await getCase('uni-app-x/app.css'),
      {
        uniAppX: true,
      },
    )
    expect(css).toMatchSnapshot('css')
  })

  it('tailwindcss3 index.uvue.ts case 1', async () => {
    const { jsHandler, styleHandler } = getCompilerContext({
      uniAppX: true,
    })
    const set = new Set<string>()
    set.add('mt-[32.43rpx]')
    set.add('bg-[#322323]')
    set.add('text-[#844343]')
    const { code } = jsHandler(
      await getCase('uni-app-x/tw3-index.uvue.ts'),
      set,
      {
        babelParserOptions: {
          sourceType: 'unambiguous',
          plugins: [
            'typescript',
          ],
        },
      },
    )
    expect(code).toMatchSnapshot()

    const { css } = await styleHandler(
      await getCase('uni-app-x/App.uvue?vue&type=style&index=0&lang.css'),
      {
        uniAppX: true,
      },
    )
    expect(css).toMatchSnapshot('css')
  })
})
