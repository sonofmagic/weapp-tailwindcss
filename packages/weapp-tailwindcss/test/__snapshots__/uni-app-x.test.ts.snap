// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uni-app-x > tailwindcss3 index.uvue.ts case 1 > css 1`] = `
"::before,::after {
    --tw-content: ""}
view,text,::before,::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgba(59, 130, 246, 0.5);
    --tw-ring-offset-shadow: 0 0 rgba(0,0,0,0);
    --tw-ring-shadow: 0 0 rgba(0,0,0,0);
    --tw-shadow: 0 0 rgba(0,0,0,0);
    --tw-shadow-colored: 0 0 rgba(0,0,0,0);
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  ;
    --tw-contain-size:  ;
    --tw-contain-layout:  ;
    --tw-contain-paint:  ;
    --tw-contain-style:  
;
    box-sizing: border-box;
    border-width: 0;
    border-style: solid;
    border-color: currentColor}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgba(59, 130, 246, 0.5);
    --tw-ring-offset-shadow: 0 0 rgba(0,0,0,0);
    --tw-ring-shadow: 0 0 rgba(0,0,0,0);
    --tw-shadow: 0 0 rgba(0,0,0,0);
    --tw-shadow-colored: 0 0 rgba(0,0,0,0);
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  ;
    --tw-contain-size:  ;
    --tw-contain-layout:  ;
    --tw-contain-paint:  ;
    --tw-contain-style:  
}
.mt-_32d43rpx_ {
    margin-top: 32.43rpx
}
.bg-_h322323_ {
    --tw-bg-opacity: 1;
    background-color: rgba(50, 35, 35, var(--tw-bg-opacity, 1))
}
.px-4 {
    padding-left: 1rem;
    padding-right: 1rem
}
.text-_h844343_ {
    --tw-text-opacity: 1;
    color: rgba(132, 67, 67, var(--tw-text-opacity, 1))
}
"
`;

exports[`uni-app-x > tailwindcss3 index.uvue.ts case 1 1`] = `
"
	const __sfc__ = defineComponent({
		data() {
			return {
				title: 'Hello'
			}
		},
		onLoad() {

		},
		methods: {

		}
	})

export default __sfc__
function GenPagesIndexIndexRender(this: InstanceType<typeof __sfc__>): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return _cE("view", _uM({ class: "px-4" }), [
    _cE("view", _uM({ class: "mt-_32d43rpx_ bg-_h322323_" }), [
      _cE("text", _uM({ class: "text-_h844343_" }), _tD(_ctx.title), 1 /* TEXT */)
    ]),
    _cE("text", _uM({ class: "icebreaker" }), "icebreaker")
  ])
}
const GenPagesIndexIndexStyles = [_uM([["icebreaker", _pS(_uM([["color", "#844343"]]))]])]
"
`;

exports[`uni-app-x > uvue.ts case 0 1`] = `
"
const __sfc__ = defineComponent({
  data() {
    return {
      title: 'Hello'
    }
  },
  onLoad() {

  },
  methods: {

  }
})

export default __sfc__
function GenPagesIndexIndexRender(this: InstanceType<typeof __sfc__>): any | null {
const _ctx = this
const _cache = this.$.renderCache
return createElementVNode("view", utsMapOf({ class: "px-4" }), [
  createElementVNode("view", utsMapOf({ class: "mt-_32d43rpx_ bg-_h322323_" }), [
    createElementVNode("text", utsMapOf({ class: "text-_h844343_" }), toDisplayString(_ctx.title), 1 /* TEXT */)
  ]),
  createElementVNode("text", utsMapOf({ class: "icebreaker" }), "icebreaker")
])
}
const GenPagesIndexIndexStyles = [utsMapOf([["icebreaker", padStyleMapOf(utsMapOf([["color", "#844343"]]))]])]
"
`;

exports[`uni-app-x > uvue.ts case 1 > css 1`] = `
"/*! tailwindcss v4.1.4 | MIT License | https://tailwindcss.com */
page,:host {
  --spacing: .25rem;
}
.mt-_32d43rpx_ {
  margin-top: 32.43rpx;
}
.bg-_h322323_ {
  background-color: #322323;
}
.px-4 {
  padding-left: calc(var(--spacing) * 4);
  padding-right: calc(var(--spacing) * 4);
}
.text-_h844343_ {
  color: #844343;
}
"
`;

exports[`uni-app-x > uvue.ts case 1 1`] = `
"
const __sfc__ = defineComponent({
  data() {
    return {
      title: 'Hello'
    }
  },
  onLoad() {

  },
  methods: {

  }
})

export default __sfc__
function GenPagesIndexIndexRender(this: InstanceType<typeof __sfc__>): any | null {
const _ctx = this
const _cache = this.$.renderCache
return createElementVNode("view", utsMapOf({ class: "px-4" }), [
  createElementVNode("view", utsMapOf({ class: "mt-_32d43rpx_ bg-_h322323_" }), [
    createElementVNode("text", utsMapOf({ class: "text-_h844343_" }), toDisplayString(_ctx.title), 1 /* TEXT */)
  ]),
  createElementVNode("text", utsMapOf({ class: "icebreaker" }), "icebreaker")
])
}
const GenPagesIndexIndexStyles = [utsMapOf([["icebreaker", padStyleMapOf(utsMapOf([["color", "#844343"]]))]])]
"
`;
